README
NAME
    Encode::UTF8Mac - "utf-8-mac" encoding, a variant utf-8 used by Mac OSX

SYNOPSIS
      use Encode;
      use Encode::UTF8Mac;
  
      my ($filename) = <*.txt>;

      $filename = Encode::decode('utf-8', $filename);
      # => "poke\x{0301}mon.txt" (NFD é)

      $filename = Encode::decode('utf-8-mac', $filename);
      # => "pok\x{00E9}mon.txt" (NFC é)

DESCRIPTION
    Encode::UTF8Mac provides a encoding called "utf-8-mac" for Mac OSX.

    On Mac OSX, utf-8 encoding is used and it is NFD (Normalization Form
    canonical Decomposition). If you want to get NFC (Normalization Form
    canonical Composition) character you need to use Unicode::Normalize's
    "NFC()".

    However, Mac OSX filesystem does not follow the exact specification.
    Specifically, the following ranges are not decomposed.

      U+2000-U+2FFF
      U+F900-U+FAFF
      U+2F800-U+2FAFF

    <http://developer.apple.com/library/mac/#qa/qa2001/qa1173.html>

    In iconv (bundled Mac), this encoding can be using as "utf-8-mac". This
    module adds "utf-8-mac" encoding for Encode, it encode/decode text with
    that rule in mind. This will help when you decode file name on Mac.

    See more specific example:
    <http://perl-users.jp/articles/advent-calendar/2010/english/24>

ENCODING
    utf-8-mac

        *   Encode::decode('utf-8-mac', $octets)

            Decode as utf-8, and normalize form C except special range using
            Unicode::Normalize.

        *   Encode::encode('utf-8-mac', $string)

            Normalize form D except special range using Unicode::Normalize,
            and encode as utf-8.

            OSX file system change NFD automatically. So actually, this is
            not necessary.

COOKBOOK
      use Encode;
      use Encode::Locale;
  
      # change locale_fs "utf-8" to "utf-8-mac"
      if ($^O eq 'darwin') {
          require Encode::UTF8Mac;
          $Encode::Locale::ENCODING_LOCALE_FS = 'utf-8-mac';
      }
  
      $filename = Encode::decode('locale_fs', $filename);

    If you are using Encode::Locale, you may want to do this.

SEE ALSO
    Encode::Locale - provides usefull "magic" encoding.

    Unicode::Normalize::Mac - contains "utf-8-mac" logic.

AUTHOR
    Naoki Tomita <tomita@cpan.org>

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Encode::UTF8Mac

{{$NEXT}}

0.04 2014-02-13T11:14:39Z
        - fixed handling of false values (by nipotan)
          https://github.com/tomill/Encode-UTF8Mac/pull/4

0.03  Fri Mar  4 00:25:56 JST 2011
        - add Unicode::Normalize::Mac (separate normalization logic from encoding).
        - update POD

0.02  Mon Dec 20 15:11:35 JST 2010
        - first CPAN release

0.01  
        - original version


lib/Unicode/Normalize/Mac.pm
package Unicode::Normalize::Mac;
use strict;
use warnings;
use Unicode::Normalize ();
use Exporter 'import';
our @EXPORT_OK = qw( NFC_mac NFD_mac );

# http://developer.apple.com/library/mac/#qa/qa2001/qa1173.html
my $decompose = qr/([^\x{2000}-\x{2FFF}\x{F900}-\x{FAFF}\x{2F800}-\x{2FAFF}]*)/;

sub NFC_mac {
    my ($unicode) = @_;
    return unless defined $unicode;
    $unicode =~ s/$decompose/Unicode::Normalize::NFC($1)/eg;
    $unicode;
}

sub NFD_mac {
    my ($unicode) = @_;
    return unless defined $unicode;
    $unicode =~ s/$decompose/Unicode::Normalize::NFD($1)/eg;
    $unicode;
}

{
    no warnings 'once';
    *NFC = \&NFC_mac;
    *NFD = \&NFD_mac;
}

1;
__END__

=encoding utf-8

=head1 NAME

Unicode::Normalize::Mac - Unicode normalization same way as Mac OSX file system

=head1 SYNOPSIS

  use Unicode::Normalize::Mac qw/NFC_mac/;
  
  my $text = NFC_mac("\x{FA1B}\x{2F872}\x{305F}\x{3099}");
  #    => "\x{FA1B}\x{2F872}\x{3060}"
  # Note: "\x{798F}\x{5BFF}\x{3060}" standard NFC

=head1 DESCRIPTION

This module provides Unicode normalization functions same as Mac OSX file system.
Specifically, the following ranges are not decomposed.

  U+2000-U+2FFF
  U+F900-U+FAFF
  U+2F800-U+2FAFF

L<http://developer.apple.com/library/mac/#qa/qa2001/qa1173.html>

=head1 FUNCTIONS

=over 4

=item NFC(), NFD()

  my $text = Unicode::Normalize::Mac::NFC($text);

Same as L<Unicode::Normalize>::NFC() / NFD(), but except some characters.

=back

=head1 EXPORT

None by default.

=over 4

=item NFC_mac(), NFD_mac()
  
  use Unicode::Normalize::Mac qw/NFC_mac/;
  my $text = NFC_mac($text);

These exportable functions are alias to Unicode::Normalize::Mac::NFC() / NFD().

=back

=head1 SEE ALSO

L<Unicode::Normalize>

L<Encode::UTF8Mac> - provides "utf-8-mac" encoding.

=head1 AUTHOR

Naoki Tomita E<lt>tomita@cpan.orgE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Encode/UTF8Mac.pm
package Encode::UTF8Mac;
use 5.008001;
use strict;
use warnings;
our $VERSION = '0.03';

use Encode ();
use Unicode::Normalize::Mac;
use base qw(Encode::Encoding);

__PACKAGE__->Define(qw(utf-8-mac));

my $utf8 = Encode::find_encoding('utf-8');

sub decode($$;$) {
    my ($self, $octets, $check) = @_;
    return unless defined $octets;
    my $string = $utf8->decode($octets, $check || Encode::FB_DEFAULT);
    $string = Unicode::Normalize::Mac::NFC($string);
}

sub encode($$;$) {
    my ($self, $string, $check) = @_;
    return unless defined $string;
    $string .= '' if ref $string;
    $string = Unicode::Normalize::Mac::NFD($string);
    $utf8->encode($string, $check || Encode::FB_DEFAULT);
}

1;
__END__

=encoding utf-8

=for stopwords utf-8-mac iconv

=head1 NAME

Encode::UTF8Mac - "utf-8-mac" a variant utf-8 used by OSX filesystem

=head1 SYNOPSIS

  use Encode;
  use Encode::UTF8Mac;
  
  # some filename from osx...
  my ($filename) = <*.txt>;
  
  # it is possible to decode by "utf-8" but...
  $filename = Encode::decode('utf-8', $filename);
  # => "poke\x{0301}mon.txt" (NFD é)
  #        ^^^^^^^^^ 2 unicode strings: "LATIN SMALL LETTER E" and "COMBINING ACUTE ACCENT"
  
  # probably you want these unicode strings.
  $filename = Encode::decode('utf-8-mac', $filename);
  # => "pok\x{00E9}mon.txt" (NFC é)
  #        ^^^^^^^^ single unicode: "LATIN SMALL LETTER E WITH ACUTE"

=head1 DESCRIPTION

Encode::UTF8Mac provides a encoding named "utf-8-mac".

On OSX, utf-8 encoding is used and it is NFD (Normalization Form
canonical Decomposition) form. If you want to get NFC (Normalization Form
canonical Composition) character you need to use L<Unicode::Normalize>'s
C<NFC()>.

However, OSX filesystem does not follow the exact specification.
Specifically, the following ranges are not decomposed.

  U+2000-U+2FFF
  U+F900-U+FAFF
  U+2F800-U+2FAFF

L<http://developer.apple.com/library/mac/#qa/qa2001/qa1173.html>

iconv (bundled Mac) can use this encoding as "utf-8-mac".

This module adds same name "utf-8-mac" encoding for L<Encode>,
it encode/decode text with that rule in mind. This will help
when you decode file name on Mac.

See more information and Japanese example:

L<Encode::UTF8Mac makes you happy while handling file names on MacOSX|http://perl-users.jp/articles/advent-calendar/2010/english/24>

=head1 ENCODING

=over 4

=item utf-8-mac

=over 4

=item * Encode::decode('utf-8-mac', $octets)

Decode as utf-8, and normalize form C except special range
using Unicode::Normalize.

=item * Encode::encode('utf-8-mac', $string)

Normalize form D except special range using Unicode::Normalize,
and encode as utf-8.

OSX file system change NFD automatically. So actually, this is not necessary.

=back

=back

=head1 COOKBOOK

  use Encode;
  use Encode::Locale;
  
  # change locale_fs "utf-8" to "utf-8-mac"
  if ($^O eq 'darwin') {
      require Encode::UTF8Mac;
      $Encode::Locale::ENCODING_LOCALE_FS = 'utf-8-mac';
  }
  
  $filename = Encode::decode('locale_fs', $filename);

If you are using L<Encode::Locale>, you may want to do this.

=head1 SEE ALSO

L<Encode::Locale> - provides useful "magic" encoding.

L<Unicode::Normalize::Mac> - this module uses it internally.

=head1 AUTHOR

Naoki Tomita E<lt>tomita@cpan.orgE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
